v2/git-initializer/main.go

```diff
@@ -0,0 +1,238 @@
+	// https://github.blog/2019-08-14-ssh-certificate-authentication-for-github-enterprise-cloud/
+	//
+	// (BRIGADE_REPO_SSH_CERT in v1)
+	// I think all we need to do is to make sure the cert file exists at /id_dsa-cert.pub
+
+	// Check for SSH Key
+	privateKey, ok := event.Project.Secrets["gitSSHKey"]
+	if ok {
+		publicKeys, err := gitssh.NewPublicKeys(
+			"git",
+			[]byte(privateKey),
+			event.Project.Secrets["gitSSHKeyPassword"],
+		)
+		if err != nil {
+			return errors.Wrapf(
+				err,
+				"error configuring authentication for remote with URL %s",
+				event.Worker.Git.CloneURL,
+			)
+		}
+		// The following avoids:
+		// "unable to find any valid known_hosts file, set SSH_KNOWN_HOSTS env variable"
+		publicKeys.HostKeyCallback = ssh.InsecureIgnoreHostKey()
+		auth = publicKeys
+	}
+
+	// Prioritize using Commit; alternatively try Ref
+	commitRef := event.Worker.Git.Commit
+	if commitRef == "" {
+		commitRef = event.Worker.Git.Ref // This will never be non-empty
+	}
+	fullRef := plumbing.NewReferenceFromStrings(commitRef, commitRef)
+	refSpec := config.RefSpec(fmt.Sprintf("+%s:%s", fullRef.Name(), fullRef.Name()))
+
+	// Initialize an empty repository with an empty working tree
+	workspace := "/var/vcs"
+	gitStorage := filesystem.NewStorage(
+		osfs.New(filepath.Join(workspace, ".git")),
+		cache.NewObjectLRUDefault(),
+	)
+
+	repo, err := git.Init(gitStorage, osfs.New(workspace))
+	if err != nil {
+		return errors.Wrapf(err, "error initializing git repository at %s", workspace)
+	}
+
+	const remoteName = "origin"
+	// If we're not dealing with an exact commit, list the remote refs
+	// to build out a full, updated refspec
+	//
+	// For example, we might be supplied with the tag "v0.1.0", but if we use
+	// this directly, we'll get: couldn't find remote ref "v0.1.0"
+	// So we need to find the full remote ref; in this case, "refs/tags/v0.1.0"
+	if gitConfig.Commit == "" {
+		// Create a new remote for the purposes of listing remote refs and finding
+		// the full ref we want
+		remoteConfig := &config.RemoteConfig{
+			Name:  remoteName,
+			URLs:  []string{gitConfig.CloneURL},
+			Fetch: []config.RefSpec{refSpec},
+		}
+		rem := git.NewRemote(gitStorage, remoteConfig)
+
+		// List remote refs
+		refs, err := rem.List(&git.ListOptions{Auth: auth})
+		if err != nil {
+			return errors.Wrap(err, "error listing remotes")
+		}
+
+		// Filter the list of refs and only keep the full ref matching our commitRef
+		var found bool
+		for _, ref := range refs {
+			// Ignore the HEAD symbolic reference
+			// e.g. [HEAD ref: refs/heads/master]
+			if ref.Type() == plumbing.SymbolicReference {
+				continue
+			}
+
+			// TODO: update this, as it is faulty... there may be multiple matches
+			// e.g. "main" might match /refs/heads/main and /refs/heads/main2
+			if strings.Contains(ref.Name().String(), fullRef.Name().String()) ||
+				strings.Contains(ref.Hash().String(), fullRef.Hash().String()) {
+				fullRef = ref
+				found = true
+			}
```

Comment:
This is the only thing I can spot that I don't think can reasonably be deferred to a follow-up.

Reply:
Ok, logic improved/made more precise.  I think we're looking good here now.  (Passes integration tests and ye olde bash tests)

Reply:
This is a major improvement!

I am wondering two things:

1. Is there still a problem due to `/refs/heads/foo` and `/refs/tags/foo` having the same short name?

1. It looks like this wouldn't allow the ref field to contain the entire long ref and still work. I feel like that has to be supported because "upstream event sources" (a GitHub web cook received by the vNext GitHub gateway, for instance) are likely to wheel and deal in long refs.

Maybe we can bypass a lot of this is we receive what's determined to be a ref in the form `/refs/heads/<branch>` or `/refs/heads/<tag>` and otherwise fall back on the above to make a "best effort" at matching?

Reply:
Ooh good point re: 1.  Yes, let me look at what we could add around cases like these.

Re: 2, full/long refs would still be supported.  Since we convert whatever is supplied to us via `Ref` into a legit `plumbing.Reference` value via `fullRef := plumbing.NewReferenceFromStrings(commitRef, commitRef)`, we then have access to the full suite of methods such as `Short()`, etc.  Unless I'm not understanding the issue correctly?

Reply:
> So, Ref comes in as /refs/heads/main and we gather the list of remote refs, where /refs/heads/main is the one we are after (exactly the same); the short names will match (main == main) and we're good.

Good up to that point.

> Well, unless there is a main tag 😄 , hence 1 above.

Right. That's where things fall apart.

I could give `/refs/heads/main` and end up pulling `/refs/tags/main`.

If I had _only_ given `main` I could understand how I ended up `/refs/tags/main`, but it would be especially worrisome to have started out being so precise and fully qualifying it as `/refs/heads/main` and still end up with the wrong thing.

Reply:
Ah, gotcha.  Ok, I've improved logic in https://github.com/brigadecore/brigade/pull/1229/commits/f9f5348d0f202e6ad12b113be821810c5aee8907 thusly:

1. Don't list remotes/look for matches if the provided ref is already a full reference
2. If we find more than one match, error out with an appropriate message

WDYT?